sharedbuffers/mapped_struct.py

```diff
@@ -106,8 +111,257 @@ def _likerobuffer(buf):
     else:
         return buffer(buf)
 
+class NONE:
+    pass
+
+@cython.cclass
+class StrongIdMap(object):
+    cython.declare(
+        __weakref__ = object,
+        preloaded = dict,
+        idmap = dict,
+        objmap = object,
+        strong_refs = object,
+        stable_set = set,
+    )
+
+    def __init__(self, strong_limit = 1 << 20 - 1, preallocate = False, strong_class = FastCache, stable_set = None):
+        """
+        Constructs a strong-referencing id map. The id map will keep strong references when necessary
+        to guarantee correct behavior, up to `strong_limit` references. After that, unused references
+        are evicted from both the strong-reference list and the id map itself, maintaining correct
+        behavior at the expense of deduplication effectiveness.
+
+        Params:
+
+            strong_limit: Keep at most this many strong references
+
+            preallocate: Passed to `strong_class` as a kwarg
+
+            strong_class: The cache constructor used for the strong referece list. By default, it uses a kind
+                of LRU cache. The constructor should accept `preallocate` as kwarg and `strong_limit` as
+                first positional argument. When given `preallocate=true`, the structure should be preallocated
+                to accommodate `strong_limit` elements. It should also accept an `eviction_callback` kwarg
+                with a callable to be called with `(key, value)` arguments when evicting items from the mapping.
+
+            stable_set: A set of shared_id s that are considered stable. That is, strong references are kept
+                by the caller, so they don't need to be tracked. This allows them to be persistent on the id map
+                and thus improve deduplication effectiveness with little overhead.
+        """
+        self.preloaded = {}
+        self.idmap = {}
+        self.strong_refs = strong_class(
+            strong_limit,
+            eviction_callback = functools.partial(
+                self._on_strong_evict,
+                weakref.ref(self),
+            ),
+            preallocate = preallocate,
+        )
+        self.objmap = weakref.WeakValueDictionary()
+        if stable_set is not None and not isinstance(stable_set, set):
+            stable_set = set(stable_set)
+        self.stable_set = stable_set
+
+    @staticmethod
+    @cython.locals(self = 'StrongIdMap')
+    def _on_strong_evict(wself, key, value):
+        self = wself()
+        if self is not None:
+            self.pop(key, None)
+
+    def __len__(self):
+        return len(self.idmap) + len(self.preloaded)
```

Comment:
is this actually accurate ? based on the iter protocol below, it appears that the actual length would be the length of the intersection between the idmap and the preloaded map. Then again, it's possible that we don't really care about this length being 100% accurate.

Reply:
Yeah, it's not accurate, but it's not worth computing the intersection to get an accurate length either, IMO. Depends on what we use the len for, currently, approximate is enough (it's for logging and stuff like that)